kernel/src/main/java/org/kframework/kast/KastFrontEnd.java

```diff
@@ -92,14 +105,51 @@ public int run() {
 
             org.kframework.kore.Sort sort = options.sort;
             if (sort == null) {
-                if (env.get("KRUN_SORT") != null) {
+                if (env.get("KRUN_SORT") != null)
                     sort = Outer.parseSort(env.get("KRUN_SORT"));
-                } else {
-                    sort = def.programStartSymbol;
+            }
+
+            if (options.input.equals(InputModes.RULE)) {
+                options.module = def.executionModule().name();
+                if (sort == null)
+                    sort = Sorts.K();
+                Module mod = def.ruleParsingModuleFor(options.module)
+                        .getOrElse(() -> {throw KEMException.innerParserError("Module " + options.module + " not found. Specify a module with -m.");});
+                String stringToParse = FileUtil.read(options.stringToParse());
+                Source source = options.source();
+
+                try (ParseInModule parseInModule = RuleGrammarGenerator.getCombinedGrammar(mod, true, null)) {
+                    Tuple2<Either<Set<KEMException>, K>, Set<KEMException>> res = parseInModule.parseString(stringToParse, sort, source);
+                    kem.addAllKException(res._2().stream().map(KEMException::getKException).collect(Collectors.toSet()));
+                    if (res._1().isLeft()) {
+                        throw res._1().left().get().iterator().next();
+                    }
+                    // important to get the extension module for unparsing because it contains generated syntax
+                    // like casts, projections and others
+                    Module unparsingMod = parseInModule.getExtensionModule();
+                    K parsed = new TreeNodesToKORE(Outer::parseSort, true).down(res._1().right().get());
+
+                    if (options.expandMacros) {
+                        parsed = ExpandMacros.forNonSentences(unparsingMod, files.get(), def.kompileOptions, false).expand(parsed);
+                    }
+                    ConfigurationInfoFromModule configInfo = new ConfigurationInfoFromModule(mod);
+                    LabelInfo labelInfo = new LabelInfoFromModule(mod);
+                    SortInfo sortInfo = SortInfo.fromModule(mod);
+
+                    Rule r = (Rule) new AddImplicitComputationCell(configInfo, labelInfo).apply(mod, Rule.apply(parsed, BooleanUtils.TRUE, BooleanUtils.TRUE, Att.empty()));
+                    r = (Rule) new ConcretizeCells(configInfo, labelInfo, sortInfo, mod, true).concretize(mod, r);
+
+                    kprint.get().prettyPrint(def.kompiledDefinition, unparsingMod, s -> kprint.get().outputFile(s), r.body(), sort);
```

Comment:
```suggestion
                    Rule r = (Rule) new AddImplicitComputationCell(configInfo, labelInfo).apply(mod, Rule.apply(parsed, BooleanUtils.TRUE, BooleanUtils.TRUE, Att.empty()));
                    r = (Rule) new ConcretizeCells(configInfo, labelInfo, sortInfo, mod, true).concretize(mod, r);

                    kprint.get().prettyPrint(def.kompiledDefinition, unparsingMod, s -> kprint.get().outputFile(s), r.body(), sort);
```

Can we make it a flag whether we do configuration completion or not? Something like `--configuration-completion`? We can default to having it turned off.

We should add a test of both uses of the flag.

Reply:
So you are ok with `<generatedTop>` cell?
What about `--sort-annotations` default to false to remove the `#SemanticCastToXXX`?

Reply:
It would be better if we could handle the user supplying `<generatedTop>` cell as well. I guess we need to make sure to include the configuration module into the grammar we use?

So you are saying that default behavior would be to remove the semantic casts, and with `--sort-annotations` you can keep them? That sounds OK, but I think it's less descriptive of what it's actually doing, which is removing semantic casts, `--no-remove-semantic-casts`. Up to you on, either works for me.